cub/warp/specializations/warp_exchange_shfl.cuh

```diff
@@ -0,0 +1,291 @@
+      // │B6│ B7 │D6│ D7 │F6│ F7 │H6│ H7
+      // └──┘    └──┘    └──┘    └──┘
+      //
+      //       RESULT
+      //       ======
+      //  t0  t1  t2  t3  t4  t5  t6  t7
+      //
+      //  A0  B0  C0  D0  E0  F0  G0  H0
+      //
+      //
+      //  A1  B1  C1  D1  E1  F1  G1  H1
+      //
+      //
+      //  A2  B2  C2  D2  E2  F2  G2  H2
+      //
+      //
+      //  A3  B3  C3  D3  E3  F3  G3  H3
+      //
+      //
+      //  A4  B4  C4  D4  E4  F4  G4  H4
+      //
+      //
+      //  A5  B5  C5  D5  E5  F5  G5  H5
+      //
+      //
+      //  A6  B6  C6  D6  E6  F6  G6  H6
+      //
+      //
+      //  A7  B7  C7  D7  E7  F7  G7  H7
+      //
+
+      // NOTE: Do *NOT* try to refactor this code to use a reference, since nvcc
+      //       tends to choke on it and then drop everything into local memory.
+      const InputT send_val = (xor_bit_set ? CompileTimeArrayImpl<OutputT, IDX>::val
+                                           : CompileTimeArrayImpl<OutputT, IDX + NUM_ENTRIES>::val);
+      const InputT recv_val = __shfl_xor_sync(mask, send_val, NUM_ENTRIES, ITEMS_PER_THREAD);
+      (xor_bit_set ? CompileTimeArrayImpl<OutputT, IDX>::val
+                   : CompileTimeArrayImpl<OutputT, IDX + NUM_ENTRIES>::val) = recv_val;
+
+      constexpr int next_idx = IDX + 1 + ((IDX + 1) % NUM_ENTRIES == 0) * NUM_ENTRIES;
+      CompileTimeArrayImpl<OutputT, next_idx>::template Foreach<NUM_ENTRIES>(xor_bit_set, mask);
+    }
+
+    // terminate recursion
+    __device__ void TransposeImpl(unsigned int, unsigned int, Int2Type<0>) {}
+
+    template <int NUM_ENTRIES>
+    __device__ void TransposeImpl(const unsigned int lane_id,
+                                  const unsigned int mask,
+                                  Int2Type<NUM_ENTRIES>)
+    {
+      const bool xor_bit_set = lane_id & NUM_ENTRIES;
+      Foreach<NUM_ENTRIES>(xor_bit_set, mask);
+
+      TransposeImpl(lane_id, mask, Int2Type<NUM_ENTRIES / 2>());
+    }
+
+  public:
+    __device__ CompileTimeArrayImpl(const InputT (&input_items)[ITEMS_PER_THREAD],
+                                    OutputT (&output_items)[ITEMS_PER_THREAD])
+        : CompileTimeArrayImpl<OutputT, IDX + 1>{input_items, output_items}
+        , val{input_items[IDX]}
+    {}
+
+    __device__ ~CompileTimeArrayImpl() { this->output_items[IDX] = val; }
+
+    __device__ void Transpose(const unsigned int lane_id, const unsigned int mask)
+    {
+      TransposeImpl(lane_id, mask, Int2Type<ITEMS_PER_THREAD / 2>());
+    }
+  };
+
+  const unsigned int lane_id;
+  const unsigned int warp_id;
+  const unsigned int member_mask;
+
+public:
+  using TempStorage = NullType;
+
+  WarpExchangeShfl() = delete;
+
+  explicit __device__ __forceinline__ WarpExchangeShfl(TempStorage &)
+      : lane_id(IS_ARCH_WARP ? LaneId() : (LaneId() % LOGICAL_WARP_THREADS))
+      , warp_id(IS_ARCH_WARP ? 0 : (LaneId() / LOGICAL_WARP_THREADS))
+      , member_mask(WarpMask<LOGICAL_WARP_THREADS>(warp_id))
+  {}
+
```

Comment:
I think this specialization should provide all the methods. Methods that do not have implementation, should have static assert with dependent false, saying that this specialization is not supporting given use case:
```cpp
template<typename T>
struct dependent_false
{
    static constexpr bool value = false;
};

template <typename OutputT>
__device__ __forceinline__ void X(const InputT (&input_items)[ITEMS_PER_THREAD],
                                  OutputT (&output_items)[ITEMS_PER_THREAD])
{
  static_assert(dependent_false<OutputT>::value, "Shuffle specialization of warp exchange does not support X");
}
```

This would produce user-friendly errors in case of the API misuse. You can omit addressing this comment in this PR if you don't have time. I can file an issue and address this later.